state-chain/pallets/cf-elections/src/electoral_systems/oracle_price/state_machine.rs

```diff
@@ -0,0 +1,541 @@
+};
+use sp_std::{
+	ops::{Index, IndexMut},
+	vec,
+};
+
+#[cfg(test)]
+use proptest_derive::Arbitrary;
+
+//--------------- configuration trait -----------------
+
+pub trait OPTypes: 'static + Sized + CommonTraits {
+	type Price: PriceTrait + CommonTraits + Ord + Default + MaybeArbitrary;
+
+	type AssetPair: AssetPairTrait + CommonTraits + Ord + Sequence + MaybeArbitrary;
+
+	type GetTime: Hook<HookTypeFor<Self, GetTimeHook>> + CommonTraits;
+
+	type SafeModeEnabledHook: Hook<HookTypeFor<Self, SafeModeEnabledHook>> + CommonTraits;
+}
+
+pub trait AssetPairTrait {
+	fn to_price_unit(&self) -> PriceUnit;
+}
+
+pub trait PriceTrait: Sized {
+	fn to_price_range(&self, range: BasisPoints) -> Option<RangeInclusive<Self>>;
+}
+
+pub struct GetTimeHook;
+impl<T: OPTypes> HookType for HookTypeFor<T, GetTimeHook> {
+	type Input = ();
+	type Output = UnixTime;
+}
+
+pub struct SafeModeEnabledHook;
+impl<T: OPTypes> HookType for HookTypeFor<T, SafeModeEnabledHook> {
+	type Input = ();
+	type Output = SafeModeStatus;
+}
+
+//---------------- the primitives ------------------
+
+def_derive! {
+	#[derive(Copy, Sequence, PartialOrd, Ord, TypeInfo)]
+	#[cfg_attr(test, derive(Arbitrary))]
+	pub enum ExternalPriceChain {
+		Solana,
+		Ethereum
+	}
+}
+
+def_derive! {
+	#[derive(TypeInfo, Copy, Default)]
+	#[cfg_attr(test, derive(Arbitrary))]
+	pub enum PriceStatus {
+		UpToDate,
+		MaybeStale,
+		#[default]
+		Stale
+	}
+}
+
+def_derive! {
+	#[derive_where(Default;)]
+	#[derive(TypeInfo)]
+	#[cfg_attr(test, derive(Arbitrary))]
+	pub struct AssetState<T: OPTypes> {
+		pub timestamp: Aggregated<UnixTime>,
+		pub price: Aggregated<T::Price>,
+		pub price_status: PriceStatus,
+		pub price_spiked: bool,
+		pub minimal_price_deviation: BasisPoints
+	}
+}
+
+impl<T: OPTypes> AssetState<T> {
+	pub fn update(&mut self, response: AssetResponse<T>) {
+		if response.timestamp.median > self.timestamp.median {
+			self.timestamp = response.timestamp;
+			self.price = response.price;
+		}
+	}
+}
+
+def_derive! {
+	#[derive(TypeInfo)]
+	#[cfg_attr(test, derive(Arbitrary))]
+	pub struct AssetResponse<T: OPTypes> {
+		pub timestamp: Aggregated<UnixTime>,
+		pub price: Aggregated<T::Price>,
+	}
+}
+
+def_derive! {
+	#[derive_where(Default;)]
+	#[derive(TypeInfo)]
+	#[cfg_attr(test, derive(Arbitrary))]
+	pub struct ExternalChainState<T: OPTypes> {
+		pub price: BTreeMap<T::AssetPair, AssetState<T>>,
+	}
+}
+
+def_derive! {
+	#[derive(TypeInfo)]
+	#[cfg_attr(test, derive(Arbitrary))]
+	pub struct ExternalChainStateVote<T: OPTypes> {
+		pub price: BTreeMap<T::AssetPair, (UnixTime, T::Price)>,
+	}
+}
+
+// This is defined here for the context, but it's used by the engines
+// to decide whether they should vote with a given oracle result.
+pub fn should_vote_for_asset<T: OPTypes>(
+	(time, price): &(UnixTime, T::Price),
+	conditions: &[VotingCondition<T>],
+) -> bool {
+	use VotingCondition::*;
+	conditions.iter().all(|condition| match condition {
```

Comment:
Shouldn't the engines vote if either condition is true?

Reply:
As discussed, in case of `UpToDate` they'll vote if the timestamp is newer AND the price moved.